packages/analytics/amplify_analytics_pinpoint_dart/test/string_database_test.dart

```diff
@@ -0,0 +1,145 @@
+        await db.addString((startValue + i).toString());
+      }
+    }
+
+    /// Reload db instance and trigger a sync with memory and indexDB storage
+    Future<void> refreshDb() async {
+      db = await IndexDbStringDb.getInstance();
+    }
+
+    void validateElements(
+      Iterable<StringDatabaseElement> readElements,
+      List<String> targetValues,
+    ) {
+      expect(readElements.length, targetValues.length);
+
+      var curCount = 0;
+      for (final element in readElements) {
+        expect(element.string, targetValues[curCount]);
+        curCount++;
+      }
+    }
+
+    Future<Iterable<StringDatabaseElement>> readAndValidateValues(
+      int readCount, {
+      int startValue = 0,
+      bool checkExactCount = true,
+    }) async {
+      final readElements = await db.getStrings(readCount);
+      if (checkExactCount) {
+        expect(readElements.length, readCount);
+      }
+      var currentIndex = startValue;
+      for (final readValue in readElements) {
+        expect(readValue.string, currentIndex.toString());
+        currentIndex++;
+      }
+      return readElements;
+    }
+
+    List<int> convertReadElementsToIds(
+      Iterable<StringDatabaseElement> readElements,
+    ) {
+      final ids = <int>[];
+      for (final element in readElements) {
+        ids.add(element.id);
+      }
+      return ids;
+    }
+
+    setUp(() async {
+      await refreshDb();
+    });
+
+    tearDown(() async {
+      await db.clearIndexDb();
+    });
+
+    test('Strings stored', () async {
+      await writeValues(10);
+      await readAndValidateValues(5);
+
+      await refreshDb();
+
+      await readAndValidateValues(5);
+    });
+
+    test('getStrings can return all stored strings', () async {
+      await writeValues(10);
+      await readAndValidateValues(100, checkExactCount: false);
+
+      await refreshDb();
+      await readAndValidateValues(100, checkExactCount: false);
+    });
+
+    test('Strings deleted', () async {
+      await writeValues(10);
+      final readElements = await readAndValidateValues(5);
+
+      final ids = convertReadElementsToIds(readElements);
+      await db.deleteStrings(ids);
+
+      await readAndValidateValues(5, startValue: 5);
+
+      await refreshDb();
+
+      await readAndValidateValues(5, startValue: 5);
+    });
+
+    test('Non sequential strings deleted', () async {
+      // Save 11 elements to test delete behavior of last saved elements
+      await writeValues(11);
+
+      final readElements = await readAndValidateValues(11);
+
+      final idsToDelete = convertReadElementsToIds(
+        readElements.where((element) {
+          return int.parse(element.string).isEven;
+        }),
+      );
+      await db.deleteStrings(idsToDelete);
+      validateElements(
+        await db.getStrings(10),
+        ['1', '3', '5', '7', '9'],
+      );
+    });
+
+    test('Concurrent writes, deletes, reads of strings', () async {
```

Comment:
I have no idea what this is testing - but awaiting writes/deletes in a for loop will make these happen serially, not concurrently.

Reply:
Successive is a better way to describe.  It's meant to test that successive operations still result in a proper state.